File: FramePySelemium/ranzhi_common.py
from time import sleep

from selenium import webdriver
from selenium.webdriver.support.select import Select


class RanzhiCommon(object):
    """
    然之系统的公共模块: 只关注功能点，不允许带数据
    1. 登录功能
    2. 退出功能
    ...
    """
    base_driver = None
    base_url = None

    APP_ADMIN_SELECTOR = "#s-menu-superadmin > button > i"
    ADMIN_ADD_MEMBER_BUTTON_SELECTOR = "#shortcutBox > div > div:nth-child(1) > div > a > h3"
    ADMIN_FRAME_SELECTOR = "#iframe-superadmin"
    ADMIN_ADD_MEMBER_EDIT_ACCOUNT = "#account"
    ADMIN_ADD_MEMBER_EDIT_REALNAME = "#realname"
    ADMIN_ADD_MEMBER_EDIT_GENDER_FEMALE = "#genderf"
    ADMIN_ADD_MEMBER_EDIT_GENDER_MALE = "#genderm"
    ADMIN_ADD_MEMBER_EDIT_DEPARTMENT = "#dept"
    ADMIN_ADD_MEMBER_EDIT_ROLE = "#role"
    ADMIN_ADD_MEMBER_EDIT_PASSWORD = "#password1"
    ADMIN_ADD_MEMBER_EDIT_PASSWORD_REPEAT = "#password2"
    ADMIN_ADD_MEMBER_EDIT_EMAIL = "#email"
    ADMIN_ADD_MEMBER_SUBMIT_SELETOR = "#submit"

    def __init__(self, driver: webdriver.Firefox, base_url

    ):
    """
    构造方法
    driver: webdriver.Firefox 指的是：
    driver这个参数的类型是 webdriver.Firefox类
    driver必须是webdriver.Firefox类实例化对象
    :param driver:
    :param base_url:
    """
    self.base_driver = driver
    self.base_url = base_url


def open_page(self, url):
    driver = self.base_driver
    driver.get(self.base_url + url)
    sleep(3)


def log_in(self, account, password):
    """
    登录 log in system with account and password
    :param account:
    :param password:
    :return:
    """
    driver = self.base_driver

    # 用户账户
    element_account = driver.find_element_by_css_selector('#account')
    element_account.clear()
    element_account.send_keys(account)

    # 用户密码
    element_password = driver.find_element_by_css_selector('#password')
    element_password.clear()
    element_password.send_keys(password)

    driver.find_element_by_css_selector('#submit').click()

    sleep(3)


def change_language(self, lang):
    """
    lang = en, 英语
    lang = zh_cn，简体
    lang = zh_tw, 繁体
    :param lang:
    :return:
    """
    driver = self.base_driver
    driver.find_element_by_css_selector("#langs > button").click()
    sleep(1)
    if lang == "en":
        lang_number = 3
    elif lang == "zh_CN":
        lang_number = 1
    elif lang == "zh_TW":
        lang_number = 2
    else:
        return "Error lang_number"

    # 拼接一个 CSS 选择器的字符串
    # 用 lang_number 拼接，选择ul的li
    selector = "#langs > ul > li:nth-child(%d) > a" % lang_number
    driver.find_element_by_css_selector(selector).click()
    sleep(3)

    element_language_button = driver.find_element_by_css_selector(
        "#langs > button")
    return element_language_button.text


def log_out(self, lang):
    """
    退出然之系统
    :return:
    """
    driver = self.base_driver

    if lang == "en":
        exit_text = "Logout"
    elif lang == "zh_CN":
        exit_text = "退出"
    elif lang == "zh_TW":
        exit_text = "退出"
    else:
        return "Error lang_number"
    # 退出系统
    driver.find_element_by_css_selector('#start').click()
    sleep(1)

    driver.find_element_by_link_text(exit_text).click()
    sleep(2)


def select_admin_app(self):
    self.base_driver.find_element_by_css_selector(self.APP_ADMIN_SELECTOR).click()
    sleep(1)


def click_add_member_button(self):
    driver = self.base_driver
    frame_admin = driver.find_element_by_css_selector(self.ADMIN_FRAME_SELECTOR)
    driver.switch_to.frame(frame_admin)

    driver.find_element_by_css_selector(self.ADMIN_ADD_MEMBER_BUTTON_SELECTOR).click()
    sleep(1)

    driver.switch_to.default_content()


def add_member_data(self, member_data):
    driver = self.base_driver

    frame_admin = driver.find_element_by_css_selector(self.ADMIN_FRAME_SELECTOR)
    driver.switch_to.frame(frame_admin)

    # 开始填写表单
    # 1. 用户名
    element_account = driver.find_element_by_css_selector(self.ADMIN_ADD_MEMBER_EDIT_ACCOUNT)
    element_account.clear()
    element_account.send_keys(member_data["account"])

    # 2. 真实姓名
    element_real_name = driver.find_element_by_css_selector(self.ADMIN_ADD_MEMBER_EDIT_REALNAME)
    element_real_name.clear()
    element_real_name.send_keys(member_data["real_name"])

    # 3. 性别
    if member_data["gender"] == 'm':
        driver.find_element_by_css_selector(self.ADMIN_ADD_MEMBER_EDIT_GENDER_MALE).click()
    elif member_data["gender"] == 'f':
        driver.find_element_by_css_selector(self.ADMIN_ADD_MEMBER_EDIT_GENDER_MALE).click()

    # 4. 部门
    # 遇到了 <select>
    # 步骤
    # 1) 找到 select 元素
    # 2) 把该元素构造成一个 Select 类的对象
    # 3) 调用对象的方法：select_by_index(): 按照下标进行选择
    element_select_dept = driver.find_element_by_css_selector(self.ADMIN_ADD_MEMBER_EDIT_DEPARTMENT)
    object_select_dept = Select(element_select_dept)
    object_select_dept.select_by_index(member_data["dept"])

    # 5 角色
    # 步骤
    # 1) 找到 select 元素
    # 2) 把该元素构造成一个 Select 类的对象
    # 3) 调用对象的方法：select_by_visible_text()： 按照显示的字符进行选择
    element_select_role = driver.find_element_by_css_selector(self.ADMIN_ADD_MEMBER_EDIT_ROLE)
    object_select_role = Select(element_select_role)
    object_select_role.select_by_index(member_data["role"])

    # 6 密码
    element_real_password = driver.find_element_by_css_selector(self.ADMIN_ADD_MEMBER_EDIT_PASSWORD)
    element_real_password.clear()
    element_real_password.send_keys(member_data["password"])

    #  7 确认密码
    element_real_password2 = driver.find_element_by_css_selector(self.ADMIN_ADD_MEMBER_EDIT_PASSWORD_REPEAT)
    element_real_password2.clear()
    element_real_password2.send_keys(member_data["password"])

    # 8 邮件
    element_real_email = driver.find_element_by_css_selector(self.ADMIN_ADD_MEMBER_EDIT_EMAIL)
    element_real_email.clear()
    element_real_email.send_keys(member_data["email"])

    sleep(1)
    # 点击 保存
    driver.find_element_by_css_selector(self.ADMIN_ADD_MEMBER_SUBMIT_SELETOR).click()

    driver.switch_to.default_content()
    sleep(5)

File: FrameSelemium/src/main/java/RanzhiCommon.java
import org.openqa.selenium.By;
import org.openqa.selenium.WebDriver;
import org.openqa.selenium.WebElement;
import org.openqa.selenium.support.ui.Select;

/**
 * Created by Linty on 2/12/2017.
 * 然之系统给的业务类
 * 作为公共的模块，可以给所有的用例使用
 */
public class RanzhiCommon {

    // 声明全局变量，每个方法都可以用下面的变量。
    private WebDriver baseDriver = null;
    private String baseUrl = null;

    // 声明常量，不可以再次被修改
    private final String APP_ADMIN_SELECTOR = "#s-menu-superadmin > button > i";
    private final String ADMIN_ADD_MEMBER_BUTTON_SELECTOR = "#shortcutBox > div > div:nth-child(1) > div > a > h3";
    private final String ADMIN_FRAME_SELECTOR = "#iframe-superadmin";
    private final String ADMIN_ADD_MEMBER_EDIT_ACCOUNT = "#account";
    private final String ADMIN_ADD_MEMBER_EDIT_REALNAME = "#realname";
    private final String ADMIN_ADD_MEMBER_EDIT_GENDER_FEMALE = "#genderf";
    private final String ADMIN_ADD_MEMBER_EDIT_GENDER_MALE = "#genderm";
    private final String ADMIN_ADD_MEMBER_EDIT_DEPARTMENT = "#dept";
    private final String ADMIN_ADD_MEMBER_EDIT_ROLE = "#role";
    private final String ADMIN_ADD_MEMBER_EDIT_PASSWORD = "#password1";
    private final String ADMIN_ADD_MEMBER_EDIT_PASSWORD_REPEAT = "#password2";
    private final String ADMIN_ADD_MEMBER_EDIT_EMAIL = "#email";
    private final String ADMIN_ADD_MEMBER_SUBMIT_SELETOR = "#submit";


    /**
     * 打开然之登录页面的业务
     */
    void openPage() throws InterruptedException {
        WebDriver driver = this.baseDriver;
        driver.get(this.baseUrl);
        Thread.sleep(2000);
    }

    /**
     * 然之的切换语言的业务
     */
    String changeLanguage(String lang) throws InterruptedException {
        WebDriver driver = this.baseDriver;
        // 点击语言按钮
        driver.findElement(By.cssSelector("#langs > button")).click();
        Thread.sleep(500);
        // 用Css Selector 选择 语言
        String langNumber = "";
        switch (lang) {
            case "en":
                langNumber = "3";
                break;
            case "zh_CN":
                langNumber = "1";
                break;
            case "zh_TW":
                langNumber = "2";
                break;
        }
        String langSelect = "#langs > ul > li:nth-child(" + langNumber + ") > a";
        driver.findElement(By.cssSelector(langSelect)).click();
        // 浏览器需要刷新，等待2秒钟
        Thread.sleep(2000);

        // 找到语言按钮的文字，返回给调用者
        String language =
                driver.findElement(By.cssSelector("#langs > button")).getText();

        return language;
    }

    /**
     * 然之的登录的业务
     */
    void logIn(String account, String password) throws InterruptedException {
        WebDriver driver = this.baseDriver;
        driver.findElement(By.id("account")).sendKeys(account);
        driver.findElement(By.id("password")).sendKeys(password);
        driver.findElement(By.id("submit")).click();
        // 点击登录按钮后，需要等待浏览器刷新
        Thread.sleep(2000);
    }

    /**
     * 构造方法
     * 把自己这个类（设计、模板）转换成对象的时候，调用
     *
     * @param driver 调用该业务用的 浏览器（司机）
     * @param url    调用该业务用的 地址
     */
    public RanzhiCommon(WebDriver driver, String url) {
        this.baseDriver = driver;
        this.baseUrl = url;
    }

    void logOut(String lang) throws InterruptedException {
        WebDriver driver = this.baseDriver;

        // 点头像（左下角）
        driver.findElement(By.cssSelector("#start > div")).click();
        Thread.sleep(500);

        // 在弹出的上下文菜单，选择文字(根据语言）
        // 1. 中文简体：退出
        // 2. 中文繁体：退出
        // 3. English：Logout
        // 用Switch来写
        String exitText = "";
        switch (lang) {
            case "en":
                exitText = "Logout";
                break;
            case "zh_CN":
            case "zh_TW":
                exitText = "退出";
                break;
        }
        driver.findElement(By.linkText(exitText)).click();
        Thread.sleep(2000);

    }


    void addContact() throws InterruptedException {
        // 点客户管理按钮
        WebDriver driver = this.baseDriver;
        driver.findElement(By.cssSelector("#s-menu-1 > button > img")).click();
        Thread.sleep(2000);

        // 点 上面 联系人 菜单
        // 这里涉及了iframe操作
        // 1. 找到 iframe 元素对应的 css selector，存到变量中
        // 2. 调用driver.switchTo().frame(刚刚被找到的iframe元素)

        WebElement elementFrame = driver.findElement(By.cssSelector("#iframe-1"));
        driver.switchTo().frame(elementFrame);

        // 进入紫禁城以后，正常操作
        driver.findElement(By.cssSelector("#mainNavbar > div.collapse.navbar-collapse > ul > li:nth-child(6) > a")).click();
        Thread.sleep(2000);

        // TODO：继续填写表单

        // 在紫荆城结束以后必须退出来
        // switchTo().frame()和 switchTo().defaultContent() 成对出现
        driver.switchTo().defaultContent();


    }


    void selectAdminApp() throws InterruptedException {
        WebDriver driver = this.baseDriver;
        // 点击齿轮按钮
        driver.findElement(By.cssSelector(this.APP_ADMIN_SELECTOR)).click();
        Thread.sleep(1000);
    }

    void clickAddMemberButton() throws InterruptedException {
        WebDriver driver = this.baseDriver;
        // 添加成员按钮 在 紫禁城中
        WebElement frameAdmin = driver.findElement(By.cssSelector(this.ADMIN_FRAME_SELECTOR));
        driver.switchTo().frame(frameAdmin);

        // 点击添加成员按钮
        driver.findElement(By.cssSelector(this.ADMIN_ADD_MEMBER_BUTTON_SELECTOR)).click();
        Thread.sleep(1000);

        // 办完事儿后，退出 紫禁城
        // 必须 进退 成对出现
        driver.switchTo().defaultContent();
    }

    void addMemberData(Member member) throws InterruptedException {
        WebDriver driver = this.baseDriver;
        // 添加成员按钮 在 紫禁城中
        WebElement frameAdmin = driver.findElement(By.cssSelector(this.ADMIN_FRAME_SELECTOR));
        driver.switchTo().frame(frameAdmin);

        // 开始填表单
        // 用户名
        WebElement elementAccount = driver.findElement(By.cssSelector(this.ADMIN_ADD_MEMBER_EDIT_ACCOUNT));
        elementAccount.clear();
        elementAccount.sendKeys(member.getAccount());

        // 真实姓名
        WebElement elementRealName = driver.findElement(By.cssSelector(this.ADMIN_ADD_MEMBER_EDIT_REALNAME));
        elementRealName.clear();
        elementRealName.sendKeys(member.getRealName());

        // 性别
        if (member.getGender() == Member.Gender.Female) {
            driver.findElement(By.cssSelector(this.ADMIN_ADD_MEMBER_EDIT_GENDER_FEMALE)).click();
        } else {
            driver.findElement(By.cssSelector(this.ADMIN_ADD_MEMBER_EDIT_GENDER_MALE)).click();
        }
        // 部门
        WebElement elementDepartment = driver.findElement(By.cssSelector(this.ADMIN_ADD_MEMBER_EDIT_DEPARTMENT));
        Select objectDepartment = new Select(elementDepartment);
        objectDepartment.selectByIndex(member.getDept());

        // 角色
        // 是个Select元素，不可以直接操作
        WebElement elementRole = driver.findElement(By.cssSelector(this.ADMIN_ADD_MEMBER_EDIT_ROLE));
        // elementRole 不可以直接操作
        // 把 elementRole 构造成一个 Select() 对象
        Select objectRole = new Select(elementRole);
        objectRole.selectByIndex(member.getRole());


        // 密码
        WebElement elementPassword = driver.findElement(By.cssSelector(this.ADMIN_ADD_MEMBER_EDIT_PASSWORD));
        elementPassword.clear();
        elementPassword.sendKeys(member.getPassword());

        // 重复密码
        WebElement elementPasswordRepeat = driver.findElement(By.cssSelector(this.ADMIN_ADD_MEMBER_EDIT_PASSWORD_REPEAT));
        elementPasswordRepeat.clear();
        elementPasswordRepeat.sendKeys(member.getPassword());

        // 邮箱
        WebElement elementEmail = driver.findElement(By.cssSelector(this.ADMIN_ADD_MEMBER_EDIT_EMAIL));
        elementEmail.clear();
        elementEmail.sendKeys(member.getEmail());

        // 点击保存按钮
        driver.findElement(By.cssSelector(this.ADMIN_ADD_MEMBER_SUBMIT_SELETOR)).click();

        Thread.sleep(5000);

        driver.switchTo().defaultContent();


    }
}
